### ISSUE TITLE
Training support in Torch Glow submodule 

### ISSUE THREAD

**Issue by dinkdeep** [pytorch/glow#4265]
Do we have some kind of multi device support also in Glow for training purposes i.e Is it possible to run different mini-batches on different devices. In the glow I don't see the separate forward pass and back ward pass specifically , Looks like the forward pass and backward pass run as  a combined integrated pass as an acyclic graph with Save nodes in the end copy back the data back to the nodes which are marked trainable(the weights Node). There seems a differentiate() function in Grad.cpp which creates the new Function with derived Nodes with SGD node (which is further broken to Mul,Add,Div), and there is a runBatch() function loops over each Batch, and reruns the Graph. Essentially its all a single device with multi instances of batch run, Are there some hooks to spread the Batches across multiple devices .

**Comment by dinkdeep** [comment]
What is the difference between the "differentiate" function present the Glow backend and the autograd automatic differentiation present in Pytorch specifically in these files
torch/csrc/distributed/autograd/engine/dist_engine.cpp
torch/csrc/autograd/engine.cpp

**Comment by jackm321** [comment]
Hi @dinkdeep 
Yes you're right, in Glow the backward and forward graphs are part of the same function and so they are integrated together. There may be some way to split training and inference using the partitioner's heterogenous partitioning functionalities but I'm not aware of anyone having tried this before.
In general there are many differences between autograd in pytorch and glow's differentiate pass, the biggest one being that glow's differentiation pass works on a static graph.

**Comment by dinkdeep** [comment]
Thanks Jack, 
Apart from that looks like there is only one optimiser SGD used by differentiate Pass in Glow vs the pytorch providing  multiple optimizers like Adagrad, Adadelta,Adam,SparseAdam, ASGD,LBFGS,RMSprop.

Looks like Glow also may need some methodology for setting up and connecting the Multi devices in different topologies like mesh, hyper cube with ranks and orders and a way to exchange/sync the gradient weights collected across the devices at the completion of backward pass across mini-batchs , Seems Pytorch  has  this Data Parallel methodology for splitting the mini Batches across the devices and then collecting and syncing the gradients using allreduce, scatter,gather communication primitives along with Distributed parallel and Model parallel methodologies supported  through  Gloo and Nccl communication library. (https://pytorch.org/docs/stable/distributed.html)

Greping through the files pf pytorch I do see they have a good support for cuda and cudnn too. 

May be be there should be some way to route torch.nn.DataParallel()  and  torch.nn.parallel.DistributedDataParallel() API's to Glow to get access to gloo/nccl C++ modules inside pytorch way it gets access to JIT IR.
